### DISCUSSION TITLE
How to send commands to Tasmota?

### DISCUSSION THREAD

**Discussion by dac2020** [arendst/Tasmota#16154] [Q&A] [answered]
I finally got tasmota onto a 4 channel relay and have been wanting to play with it.
However, no matter what I do, I can only get one thing to reply. I've tried form a browser and CLI using curl.

These are all commands I found online, only one has worked.

curl http://192.168.254.200/cm?cmnd=Power
{"Command":"Error"}[pc]# 

curl -s http://192.168.254.200/cm?cmnd=STATUS+8
curl http://192.168.254.200/cm?cmnd=POWER+TOGGLE
curl http://192.168.254.200/cm?cmnd=POWER
curl http://192.168.254.200/cm?cmnd=power
curl http://192.168.254.200/cm?cmnd=power%20on
curl http://192.168.254.200/cm?cmnd=Power %2$s
curl http://192.168.254.200/cm?cmnd=status

$ curl http://192.168.254.200/cm?cmnd=status
{"Status":{"Module":1,"DeviceName":"relay4-01","FriendlyName":["relay4-01"],"Topic":"tasmota_F8DC4C","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0}}

That's the only thing I can get to respond. I've looked at 
https://tasmota.github.io/docs/Getting-Started/#after-configuration
and
https://tasmota.github.io/docs/Commands/#wi-fi

I must be missing something really simple.

**Comment by sfromis** [comment, marked as answer]
Read the Tasmota console to see what happens, with `Weblog 3` for details. And first try out your commands in the console before even trying to send them from the outside.

Generally, using http is not the best way, Tasmota is built for being controlled via MQTT.

The `Status` command output does not look like you configured Tasmota to match the 4-relay hardware. Check this first.

**Reply by dac2020** [reply]
All I've done so far is to install tasmota on the device. I've not found any information on what I need to do in order to configure it for relays etc. I did try from two consoles first, then browser, then curl.

I have weblog set to 3 now so will try a few commands.

**Reply by dac2020** [reply]
Basically, I've been trying to find information from the board, so I can know what it calls its relays etc so I can try controlling them. It's responding to commands but no idea what the relays are called or how to control them yet.

**Reply by sfromis** [reply]
To configure your device, find the appropriate template in https://templates.blakadder.com/ and perform the steps to apply it.

Better get used to Tasmota using the web UI before going on to other avenues for trying to control it. The required commands are the same, and the Tasmota web console is preferable for getting feedback on what happens.

**Reply by dac2020** [reply]
Thank you. I'll have a look at that and follow your advise. I'm sure once I see things working, it will start coming together in terms of understanding. Just need to get the basics first then expand on that.

**Reply by dac2020** [reply]
Maybe I'm too old for this. After setting the template and looking at the many commands, I can still not find any idea what the relays are called so I can switch them on/off, etc.

Even setting the ntp server doesn't seem to be working for me.
I tried setting the two local dns server and up to three ntp servers but it's not working.
(BTW, I tried using the quote option but it just makes one single blob of text.

01:28:18.946 CMD: EthDnsServer1 192.168.1.29
01:28:18.951 SRC: WebConsole from 192.168.1.205
01:28:18.955 CMD: Grp 0, Cmd 'ETHDNSSERVER', Idx 1, Len 12, Pld 192, Data '192.168.1.29'
01:28:18.967 RSL: RESULT = {"EthDNSServer1":"192.168.1.29"}
01:28:23.434 CMD: EthDnsServer2 192.168.1.30
01:28:23.437 SRC: WebConsole from 192.168.1.205
01:28:23.442 CMD: Grp 0, Cmd 'ETHDNSSERVER', Idx 2, Len 12, Pld 192, Data '192.168.1.30'
01:28:23.450 RSL: RESULT = {"EthDNSServer2":"192.168.1.30"}
01:28:42.935 DNS: Unable to resolve '0.pool.ntp.org'
01:28:42.999 CMD: NtpServer1 0.pool.ntp.org
01:28:43.004 SRC: WebConsole from 192.168.1.205
01:28:43.007 CMD: Grp 0, Cmd 'NTPSERVER', Idx 1, Len 14, Pld 0, Data '0.pool.ntp.org'
01:28:43.011 CFG: CR 375/699, Busy 0
01:28:43.014 CFG: CR 375/699, Busy 0
01:28:43.019 RSL: RESULT = {"NtpServer1":"0.pool.ntp.org"}
01:28:47.038 DNS: Unable to resolve '0.pool.ntp.org'
01:28:54.965 CMD: NtpServer2 1.pool.ntp.org
01:28:54.968 SRC: WebConsole from 192.168.1.205
01:28:54.973 CMD: Grp 0, Cmd 'NTPSERVER', Idx 2, Len 14, Pld 1, Data '1.pool.ntp.org'
01:28:54.976 CFG: CR 375/699, Busy 0
01:28:54.979 CFG: CR 375/699, Busy 0
01:28:54.983 RSL: RESULT = {"NtpServer2":"1.pool.ntp.org"}
01:28:58.140 DNS: Unable to resolve '1.pool.ntp.org'
01:29:42.276 DNS: Unable to resolve '2.pool.ntp.org'
01:30:33.389 RSL: STATE = {"Time":"1970-01-02T01:30:33","Uptime":"1T00:35:12","UptimeSec":88512,"Heap":156,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Berry":{"HeapUsed":3,"Objects":40},"Wifi":{"AP":1,"SSId":"gst2","BSSId":"A4:2B:B0:FA:3C:AB","Channel":11,"Mode":"11n","RSSI":88,"Signal":-56,"LinkCount":1,"Downtime":"0T00:00:05"}}
01:30:33.409 RSL: SENSOR = {"Time":"1970-01-02T01:30:33","ESP32":{"Temperature":43.9},"TempUnit":"C"}
01:30:42.449 DNS: Unable to resolve '1.pool.ntp.org'

If I'm understanding some of the things I am reading, I need to also create buttons and toggles. It sure is hard to find much information on this somewhat secret project :). It seems to be mostly for developers and not end users.

**Reply by sfromis** [reply]
Do you get to the point of having the relays visible and controllable on the Tasmota web UI front page? Standard setup is that Tasmota will automatically create toggle buttons and statuses for relays. You do not need to do anything to set up such details. Only start to look at other ways of controlling relays if you have the basic automatic setup working.

Since you have Wifi connected, you should not use `Ethxxxx` commands to try to control anything at all, those are only if you have a Ethernet wired connection to the board. Thus, use the [IPAddress](https://tasmota.github.io/docs/Commands/#ipaddress) command if you *have to* manually config DNS, but the recommended preferred way is to use DHCP, and let your AP/router give Tasmota networking details like DNS server.

Tasmota tends to work "less well" without connection to a NTP server (or other real time source)

To quote text blocks, have a triple backtick line before and after the block, unlike single phrases only wrapped in single backtick.

**Reply by dac2020** [reply]
After using the template code, nothing shows up, all GPIO are still set to none. I entered it from the console. Maybe I need to enter it from somewhere else?

The device does get DHCP with dns from a DHCP server but for some reason, it's not able to resolve the NTP servers. Other things on the same network can so I'm not sure what is happening.

I received this unit last night.
https://www.aliexpress.com/item/3256802966722010.html?spm=a2g0o.order_list.0.0.117c1802DRAt6Y

This is already running tasmota and it works right out the box. Too bad the board is not identical otherwise I could have simply copied the settings. I tried just for the fun of it and when I do, the unit starts responding slowly and eventually loses it's wifi config and I have to start over.

So far, I've not been able to find any information to try and set up the config on this device but still looking.

I'll try the quote text suggestion next time, thank you.

**Reply by dac2020** [reply]
LOL, digging all over the Internet, youtube and still never finding something that shows what parameters I need to set.

**Reply by barbudor** [reply]
There's only 1 place to go :  tasmota.github.io/docs/
This is up to date and maintained by Tasmota contributors
All the rest is generally obsolete once the sun has set

If you have the template there's 2 ways:
- web gui
Go to Configuration->Others
Enter the template here
Click the Activate box and save

- console
Enter
```
backlog template { whatever is the template} ; module 0
```

Where did you got the template from? (link please)

**Reply by dac2020** [reply]
Thank you. That did the trick.
I had entered the template from the console but doing it the way you suggested worked.

'''
00:00:06.886 CMD: backlog template { whatever is the template} ; module 0
00:00:06.946 RSL: RESULT = {"Template":"Invalid JSON"}
00:00:06.957 RSL: INFO1 = {"Info1":{"Module":"T-Relay ESP32","Version":"12.0.2(tasmota)","FallbackTopic":"cmnd/DVES_F8DC4C_fb/","GroupTopic":"cmnd/tasmotas/"}}
00:00:06.972 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-F8DC4C-7244","IPAddress":"192.168.254.200"}}
00:00:06.977 RSL: INFO3 = {"Info3":{"RestartReason":"Software reset CPU","BootCount":10}}
00:00:06.991 RSL: RESULT = {"POWER1":"OFF"}
00:00:06.993 RSL: POWER1 = OFF
00:00:06.995 RSL: RESULT = {"POWER2":"OFF"}
00:00:06.997 RSL: POWER2 = OFF
00:00:06.999 RSL: RESULT = {"POWER3":"OFF"}
00:00:07.001 RSL: POWER3 = OFF
00:00:07.004 RSL: RESULT = {"POWER4":"OFF"}
00:00:07.016 RSL: POWER4 = OFF
00:00:10.072 RSL: RESULT = {"Module":{"0":"T-Relay ESP32"}}
00:00:13.310 QPC: Reset
00:00:19.532 RSL: STATE = {"Time":"1970-01-01T00:00:19","Uptime":"0T00:00:10","UptimeSec":10,"Heap":159,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":1432,"MqttCount":0,"Berry":{"HeapUsed":3,"Objects":45},"POWER1":"OFF","POWER2":"OFF","POWER3":"OFF","POWER4":"OFF","Wifi":{"AP":1,"SSId":"gst2","BSSId":"A4:2B:B0:FA:3C:AB","Channel":11,"Mode":"11n","RSSI":80,"Signal":-60,"LinkCount":1,"Downtime":"0T00:00:04"}}
00:00:19.574 RSL: SENSOR = {"Time":"1970-01-01T00:00:19","ESP32":{"Temperature":43.3},"TempUnit":"C"}
'''
Nice, I can finally play with this a little now. 

Can someone share a url where someone is discovering the relationship between the hardware, relays, LED's, connectors, etc on the board to match with the GPIOs. I just want to see how they go about that.

**Reply by barbudor** [reply]
It didn't came to you that `{ what ever is your template }` has to be replaced by your actual template?

**Reply by barbudor** [reply]
Not sure what you are looking for but a Generic guide is here https://tasmota.github.io/docs/Configuration-Procedure-for-New-Devices/

Devices such as the Robodyn generally have documentation on the manufacturer site if no-one has already created the template

**Reply by dac2020** [reply]
I'm not sure what you are saying 'it didn't come to you'. 
Yes, I did replace it as explained above but I did it from the command line, not the GUI so maybe that's where I went wrong. 

I imagine creating a template is around the same process of discovering any other hardware, like how it's done to use openwrt on a device for example. Someone has to take the hardware, examine it, try to gain access, figure out how things are connected etc. Just wondered. 

Anyways, it looks like I'm finally up and running with two devices now so will play with that. I saw somewhere that there is an open source phone app to control devices which is truly why I wanted to learn about this, to get away from all these 'custom' apps and have one single point of control for everything. 

Off to dig some more. Thanks to everyone that helped a new guy along.

**Comment by reimerp** [comment]
Most commands want Post / Urlencoded request that you can send this way:
`curl "http://tasmota-device/cm?user=admin&password=xxxx" --data-urlencode 'cmnd=Status 1'`
if you don't have set _WebPassword_ just remove the ?user stuff

**Comment by reimerp** [comment]
You can also use complex mult line commands, using the here feature (only Unix) with recent Gnu curl but this is not a tasmota issue/feature
```
    curl --silent "$URL" --data-urlencode cmnd@- <<- "F_EOF"
        Rule1
                ON Switch1#State=2 DO Event Power=%value% ENDON
                ON Switch1#State=3 DO Backlog Power1 ON; RuleTimer1 0 BREAK
                ON Event#Power DO Backlog Power1 %value%; RuleTimer1 120 BREAK
                ON Rules#Timer=1 DO Power1 OFF BREAK
F_EOF
```

**Comment by dac2020** [comment]
I've set up an MQTT server so am now using mainly mosquitto_pub and direct connections to the devices. 

Thanks for all of the input!
